balancer/rls/internal/config.go

```diff
@@ -0,0 +1,287 @@
+	if rpStrategy == rlspb.RouteLookupConfig_STRATEGY_UNSPECIFIED {
+		return nil, fmt.Errorf("rls: request_processing_strategy cannot be left unspecified in service config {%+v}", string(c))
+	}
+	defaultTarget := rlsProto.GetDefaultTarget()
+	if (rpStrategy == rlspb.RouteLookupConfig_SYNC_LOOKUP_DEFAULT_TARGET_ON_ERROR ||
+		rpStrategy == rlspb.RouteLookupConfig_ASYNC_LOOKUP_DEFAULT_TARGET_ON_MISS) && defaultTarget == "" {
+		return nil, fmt.Errorf("rls: request_processing_strategy is %s, but default_target is not set", rpStrategy.String())
+	}
+
+	if childPolicy == nil {
+		return nil, fmt.Errorf("rls: childPolicy is invalid in service config {%+v}", string(c))
+	}
+	if cfgJSON.ChildPolicyConfigTargetFieldName == "" {
+		return nil, fmt.Errorf("rls: childPolicyConfigTargetFieldName field is not set in service config {%+v}", string(c))
+	}
+	cpCfg, err := validateChildPolicyConfig(childPolicy, cfgJSON.ChildPolicyConfigTargetFieldName)
+	if err != nil {
+		return nil, err
+	}
+
+	return &lbConfig{
+		kbMap:                kbMap,
+		lookupService:        parsedTarget,
+		lookupServiceTimeout: lookupServiceTimeout,
+		maxAge:               maxAge,
+		staleAge:             staleAge,
+		cacheSizeBytes:       cacheSizeBytes,
+		rpStrategy:           rpStrategy,
+		cpName:               childPolicy.Name,
+		cpTargetField:        cfgJSON.ChildPolicyConfigTargetFieldName,
+		cpConfig:             cpCfg,
+	}, nil
+}
+
+// validateChildPolicyConfig validates the child policy config received in the
+// service config. This makes it possible for us to reject service configs
+// which contain invalid child policy configs which we know will fail for sure.
+//
+// It does the following:
+// * Unmarshals the provided child policy config into a map of string to
+//   json.RawMessage. This allows us to add an entry to the map corresponding
+//   to the targetFieldName that we received in the service config.
+// * Marshals the map back into JSON, finds the config parser associated with
+//   the child policy and asks it to validate the config.
+// * If the validation succeeded, removes the dummy entry from the map and
+//   returns it. If any of the above steps failed, it returns an error.
+func validateChildPolicyConfig(cp *loadBalancingConfig, cpTargetField string) (map[string]json.RawMessage, error) {
+	var childConfig map[string]json.RawMessage
+	if err := json.Unmarshal(cp.Config, &childConfig); err != nil {
+		return nil, fmt.Errorf("rls: json unmarshal failed for child policy config {%+v}: %v", cp.Config, err)
+	}
+	childConfig[cpTargetField], _ = json.Marshal(dummyChildPolicyTarget)
+
+	jsonCfg, err := json.Marshal(childConfig)
+	if err != nil {
+		return nil, fmt.Errorf("rls: json marshal failed for child policy config {%+v}: %v", childConfig, err)
+	}
+	builder := balancer.Get(cp.Name)
+	if builder == nil {
+		// This should never happen since we already made sure that the child
+		// policy name mentioned in the service config is a valid one.
```

Comment:
Maybe we should store the child balancer.Builder itself (_instead_ of the name? or in addition to if necessary, although the `Name` method will give it to us if we need it for debugging) so we can be 100% sure it's impossible?

We could also validate it implements `ConfigParser` at that time, and store it in the config as `type interface BalancerConfigParser { balancer.Builder; balancer.ConfigParser }`.  (Maybe this is too crazy?)

...

I see now that this creates the `lbConfig` _after_ doing validation of the child policy, so the above might sound a little strange.  However, we'll soon have code that calls the child's `ParseConfig` to retrieve its updated config with the target name set, at which point that will make more sense.

Reply:
Makes sense. Added a couple of TODOs for this.